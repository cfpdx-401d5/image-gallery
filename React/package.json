{
  "name": "image-gallery",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "body-parser": "^1.17.1",
    "cors": "^2.8.1",
    "express": "^4.15.2",
    "mongoose": "^4.8.7",
    "morgan": "^1.8.1",
    "react": "^15.4.2",
    "react-dom": "^15.4.2"
  },
  "devDependencies": {
    "chai": "^3.5.0",
    "chai-http": "^3.0.0",
    "enzyme": "^2.7.1",
    "enzyme-to-json": "^1.5.0",
    "eslint": "^3.17.1",
    "eslint-plugin-react": "^6.10.0",
    "mocha": "^3.2.0",
    "nodemon": "^1.11.0",
    "react-addons-test-utils": "^15.4.2",
    "react-scripts": "0.9.4"
  },
  "scripts": {
    "start": "react-scripts start",
    "start-server": "nodemon server.js",
    "build": "react-scripts build",
    "test": "react-scripts test --env=jsdom",
    "eject": "react-scripts eject"
  }
}
