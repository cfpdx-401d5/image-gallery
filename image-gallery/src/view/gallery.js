import React from 'react';

export default function Gallery(props) {
    const galleryObj = props.array.map((schnoodle, index) => {
        return (
            <div className="clearfix" key={index}>
                <li key={index}>
                    <p>Title: {schnoodle.title}</p>
                    <img className='gallery-img' src={schnoodle.url} alt='' />
                    <p>description: {schnoodle.description}</p>
                </li>
            </div>
        )
    });
    return (
        <ul>{galleryObj}</ul>
    )
}

Gallery.propTypes = {
    array: React.PropTypes.array
};
