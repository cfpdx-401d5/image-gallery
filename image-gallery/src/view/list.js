import React from 'react';

function ListDisplay(props) {
    return (
        <div className="clearfix">
          <li>
            <p>title: {props.schnoodle.title}</p>
            <p>description: {props.schnoodle.description}</p>
            <p>url: <a href={props.schnoodle.url}>{props.schnoodle.url}</a></p>
            <button onClick={() => props.onDelete(props.schnoodle.id)}>Delete</button>
          </li>
        </div>
    );
}

export default function List(props) {
    const listItem = props.array.map((schnoodle, index) => {
        return (
            <ListDisplay key={index} schnoodle={schnoodle} onDelete={props.onDelete}/>
        );
    });
    return (
      <ul>{listItem}</ul>
    );
}

List.propTypes = {
    array: React.PropTypes.array,
    onDelete: React.PropTypes.func
};

ListDisplay.propTypes = {
    schnoodle: React.PropTypes.object,
    onDelete: React.PropTypes.func
};

