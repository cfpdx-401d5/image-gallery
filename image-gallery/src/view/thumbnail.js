import React from 'react';

export default function Thumbnail(props) {
  const thumbnail = props.array.map((schnoodle, index) => {
    return (
      <div className="clearfix" key={index}>
        <li key={index}>
          <p height='20px' width='200'>title: {schnoodle.title}</p>
          <img className='thumbnail' src={schnoodle.url} alt='' height='100' width='100' />
        </li>
      </div>
    );
  });
  return (
    <ul>{thumbnail}</ul>
  );
}

Thumbnail.propTypes = {
  array: React.PropTypes.array
};
