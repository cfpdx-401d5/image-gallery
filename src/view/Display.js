import React from 'react';
import EditForm from '../forms/EditForm';

function ChooseView(props) {
    switch (props.view) {

        case 'gallery':
            return (
                <div>
                    <p><b>Title:</b> {props.item.title}</p>
                    <img className='gallery' src={props.item.url} alt='cute bunnys'/>
                    <p><b>Description:</b> {props.item.description}</p>
                </div>
            );

        case 'details':
            return (
                <div>
                    <p><b>Title:</b> {props.item.title}</p>
                    <p><b>Description:</b> {props.item.description}</p>
                    <p><b>URL:</b> <a href={props.item.url}>{props.item.url}</a></p>
                </div>
            );

        case 'thumbnail':
            return (
                <div>
                    <p><b>Title:</b> {props.item.title}</p>
                    <img className='thumbnail' src={props.item.url} alt='cute bunnys'/>
                </div>
            );

        default:
            return (
                'Something went wrong.'
            );
    }
}

function Buttons(props) {
    return (
        <div>
            <button className='buttons' onClick={() => props.onDelete(props.item.id)}>Delete</button>
            <button className='buttons' onClick={() => props.edit()}>Edit</button>
            <div>{props.isEditable && <EditForm item={props.item} onChange={props.onChange} onDisplay={props.edit} onEdit={props.onEdit} />}</div>
        </div>
    );
}

class Display extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            isEditable: false
        };
        this.editForm = this.editForm.bind(this);
    }

    editForm() {
        this.setState({
            isEditable: !this.state.isEditable
        });
    }

    render() {
        return (
            <li>
                <ChooseView 
                    item={this.props.item} 
                    view={this.props.view} />

                <Buttons 
                    edit={this.editForm} 
                    isEditable={this.state.isEditable}
                    onDelete={this.props.onDelete} 
                    onChange={this.props.onChange} 
                    onEdit={this.props.onEdit}
                    item={this.props.item} />
                    <hr />
            </li>
        );
    }
}

function DisplayList(props) {
    const listItem = props.array.map((item, index) => {
        return (
            <Display 
            key={index} 
            item={item} 
            onDelete={props.onDelete} 
            onChange={props.onChange} 
            onEdit={props.onEdit} 
            view={props.view} />
        );
    });
    return (
      <ul>{listItem}</ul>
    );
}

export { DisplayList };

DisplayList.propTypes = {
    array: React.PropTypes.array,
    view: React.PropTypes.string
};

Display.propTypes = {
    onDelete: React.PropTypes.func,
    onEdit: React.PropTypes.func,
    onChange: React.PropTypes.func,
    item: React.PropTypes.object,
    view: React.PropTypes.string
};

Buttons.propTypes = {
    onDelete: React.PropTypes.func,
    onEdit: React.PropTypes.func,
    onChange: React.PropTypes.func,
    item: React.PropTypes.object,
    editForm: React.PropTypes.func,
    isEditable: React.PropTypes.bool,
    edit: React.PropTypes.func
};

ChooseView.propTypes = {
    view: React.PropTypes.string,
    item: React.PropTypes.object
};
