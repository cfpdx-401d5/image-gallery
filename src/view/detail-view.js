import React from 'react';
import EditForm from '../forms/edit-form';

class DetailDisplay extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            isEditable: false
        };
        this.editForm = this.editForm.bind(this);
    }

    editForm() {
        this.setState({
            isEditable: !this.state.isEditable
        });
    }

    render() {
        return (
          <li>
            <p><b>Title:</b> {this.props.item.title}</p>
            <p><b>Description:</b> {this.props.item.description}</p>
            <p><b>URL:</b> <a href={this.props.item.url}>{this.props.item.url}</a></p>
            <button className='buttons' onClick={() => this.props.onDelete(this.props.item.id)}>Delete</button>
            <button className='buttons' onClick={() => this.editForm()}>Edit</button>
            <div>{this.state.isEditable && <EditForm item={this.props.item} onChange={this.props.onChange} onDisplay={this.editForm} onEdit={this.props.onEdit} />}</div>
            <hr />
          </li>
        );
    }
}

export default function DetailList(props) {
    const listItem = props.array.map((item, index) => {
        return (
            <DetailDisplay key={index} item={item} onDelete={props.onDelete} onChange={props.onChange} onEdit={props.onEdit} />
        );
    });
    return (
      <ul>{listItem}</ul>
    );
}

DetailList.propTypes = {
    array: React.PropTypes.array,
    onDelete: React.PropTypes.func,
    onEdit: React.PropTypes.func
};

DetailDisplay.propTypes = {
    item: React.PropTypes.object,
    onDelete: React.PropTypes.func,
    onDisplay: React.PropTypes.func,
    onChange: React.PropTypes.func,
    onEdit: React.PropTypes.func
};
