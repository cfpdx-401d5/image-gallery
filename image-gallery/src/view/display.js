import React from 'react';
import { List } from './list';
import { Gallery } from './gallery';
import { Thumbnail } from './thumbnail'
import AddSchnoodle from '../forms/form';

import puppies from '../schnoodle';

const GALLERY_TYPE = 'gallery';
const LIST_TYPE = 'list';
const THUMBNAIL_TYPE = 'thumbnail';
const ADDSCHNOODLE_TYPE = 'addSchnoodle';

function ViewSelector(props) {
    return (
        <div className="button-holder">
            <button onClick={() => props.onClick(GALLERY_TYPE)} className="action-button shadow animate orange">Gallery View</button>
            <button onClick={() => props.onClick(LIST_TYPE)} className="action-button shadow animate yellow">List View</button>
            <button onClick={() => props.onClick(THUMBNAIL_TYPE)} className="action-button shadow animate blue">Thumbnail View</button>
            <button onClick={() => props.onClick(ADDSCHNOODLE_TYPE)} className="action-button shadow animate purple">Add Schnoodle</button>
        </div>
    );
}

export default class ImageView extends React.Component {
    constructor(props) {
        super(props);
        this.handleDelete = this.handleDelete.bind(this);
        this.handleSchnoodleChange = this.handleSchnoodleChange.bind(this);
        this.state = {
            puppies: puppies()
        };
    }

    handleDelete(value) {
        const newSchnoodleArray = this.state.puppies.filter(schnoodle => { //eslint-disable-line
            if (schnoodle.id !== value) {
                return schnoodle;
            }
        });
        this.setState({
            puppies: newSchnoodleArray
        });
    }

    handleSchnoodleChange(value) {
        const newSchnoodleArray = this.state.schnoodles.concat(value);
        this.setState({
            puppies: newSchnoodleArray
        });
    }


    render() {
        const outputs = {
            [LIST_TYPE]: List,
            [GALLERY_TYPE]: Gallery,
            [THUMBNAIL_TYPE]: Thumbnail,
            [ADDSCHNOODLE_TYPE]: AddSchnoodle
        };

        const OutputComponent = outputs[this.props.view];

        return (
            <div>
                {OutputComponent
                    ? <OutputComponent array={this.state.puppies} onDelete={this.handleDelete} onChange={this.handleSchnoodleChange}/>
                    : <text className="view">Click a button to view schnoodles:</text>
                }
                <ViewSelector onClick={this.props.onClick} />
            </div>
        );
    }
}

ImageView.constructor.propTypes = {
    view: React.PropTypes.string,
    state: React.PropTypes.object,
    onClick: React.PropTypes.func
};

ViewSelector.propTypes = {
    onClick: React.PropTypes.func
};
