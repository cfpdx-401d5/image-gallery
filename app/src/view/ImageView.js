import React from 'react';
import { DisplayList } from './Display';
import BunnyForm from '../forms/BunnyForm';
const constants = require('../constants');

function fetcher(options) {
    const { method, path, body } = options;
    return fetch(`http://localhost:4000${path}`, {
        method: method,
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(body),
    });
}

function ViewSelector(props) {
    const viewTypes = Object.values(constants);
    const viewButtons = viewTypes.map((item, index) => {
        return <button className='buttons' key={index} onClick={() => props.onClick(item)}>{item} view</button>;
    });
    return (
       <div className='main-buttons'>
        { viewButtons }
      </div>
    );
}

export default class ImageView extends React.Component {
    constructor(props) {
        super(props);
        this.deleteBunny = this.deleteBunny.bind(this);
        this.addBunny = this.addBunny.bind(this);
        this.editBunny = this.editBunny.bind(this);
        this.state = {
            bunnies: null
        };
    }

    //TODO: Refactor all change handlers to one
    componentDidMount() {
        this.getBunnies();
    }

    getBunnies() {
        fetcher({
            method: 'GET',
            path: '/'
        })
        .then(res => {
            return res.json();
        })
        .then(res => {
            this.setState({
                bunnies: res
            });
        })
        .catch(err => {
            console.error('err: ', err);
        });
    }

    addBunny(value) {
        fetcher({
            method: 'POST',
            path: '/',
            body: value
        })
        .then(res => {
            return res.json();
        })
        .then(res => {
            const newBunnyArray = this.state.bunnies.concat(res);
            this.setState({
                bunnies: newBunnyArray
            });
        })
        .catch(err => {
            console.error('err: ', err);
        });
    }

    deleteBunny(value) {
        fetcher({
            method: 'DELETE',
            path: `/${value}`
        })
        .then(res => {
            return res.json();
        })
        .then(res => {
            if (res.deleted) {
                const newBunnyArray = this.state.bunnies.filter(item => { //eslint-disable-line
                    if (item._id !== value) {
                        return item;
                    }
                });
                this.setState({
                    bunnies: newBunnyArray
                });
            }        
        })
        .catch(err => {
            console.error('err: ', err);
        });
    }

    editBunny(value) {
        fetcher({
            method: 'PUT',
            path: `/${value.id}`,
            body: {
                title: value.title,
                description: value.description,
                url: value.url
            }
        })
        .then(res => {
            return res.json();
        })
        .then(res => {
            this.state.bunnies.forEach((item, index) => {
                if (value.id === item._id) {
                    let newBunnyArray = this.state.bunnies;
                    newBunnyArray[index] = value;
                    this.setState({ bunnies: newBunnyArray });
                } 
            });
        })
        .catch(err => {
            console.error('err: ', err);
        });
    }

    render() {
        const outputs = {
            [constants.THUMBNAIL_TYPE]: DisplayList,
            [constants.DETAILS_TYPE]: DisplayList,
            [constants.GALLERY_TYPE]: DisplayList,
            [constants.BUNNYFORM_TYPE]: BunnyForm
        };    
    
        const OutputComponent = outputs[this.props.view];

        return (
            <div>
                <ViewSelector onClick={this.props.onClick}/>
                {OutputComponent 
                    ? <OutputComponent 
                        view={this.props.view}
                        array={this.state.bunnies} 
                        onDelete={this.deleteBunny} 
                        onChange={this.addBunny} 
                        onEdit={this.editBunny} />
                    : <h3>Click a button to view bunnies!</h3>
                }          
            </div>
        );
    }
}

ViewSelector.propTypes = {
    onClick: React.PropTypes.func
};

ImageView.constructor.propTypes = {
    view: React.PropTypes.string,
    state: React.PropTypes.object,
    onClick: React.PropTypes.func
};
