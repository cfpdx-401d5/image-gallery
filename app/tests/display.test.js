/* eslint-disable */
import React from 'react';
import {shallow, render} from 'enzyme';
import toJson from 'enzyme-to-json';
import { DisplayList } from '../src/view/Display';

describe('display', () => {
    it('display', () => {
        const rendered = shallow(<DisplayList array={[]} />);
        expect(toJson(rendered)).toMatchSnapshot();
    });
});