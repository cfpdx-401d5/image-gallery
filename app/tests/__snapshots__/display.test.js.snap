exports[`display display 1`] = `<ul />`;
